nixos/modules/services/networking/wireguard.nix

```diff
@@ -160,6 +160,14 @@ let
         interval of 25 seconds; however, most users will not need this.'';
       };
 
+      table = mkOption {
+        default = "main";
+        type = types.str;
+        description = ''The kernel routing table to add this peer's associated
```

Comment:
Interesting.

If used as described in the [Conceptual Overview](https://www.wireguard.com/#conceptual-overview), WireGuard implements (something similar to) *policy-based VPN*. That is, a WireGuard interface abstracts over a multitude of peer associations, and each peer can be associated with many other peers. Routing decisions are made only within WireGuard code, and ingress filtering is performed automatically.

What you can also do is use WireGuard interfaces as "dumb" point-to-point links, with `AllowedIPs` set to `0/0` on both sides, and make all routing decisions the same way they are made with other types of interfaces. This is what one would call *route-based VPN*. Instead of maintaining `AllowedIPs`, one would need to maintain sensible static routes. Or you could run dynamic routing protocols across WireGuard links. You'd still need to do ingress filtering (BCP-84-like) yourself.

In this latter case, you are right: inserting all routes associated with a single interface into a single routing table is completely sufficient.

In the former case, however, using different tables for different routes on a single interface is useful. Suppose you configure a WireGuard interface with, say, five different peers in different networks, on one of your machines. Furthermore, let's say you run five VMs, with a TAP interface each, on the same machine. Normally, the associated `qemu` processes all get their packets routed using the same routing table (`main`). You could, however, use `ip rule add iif ${TAPINTERFACE} table ${TABLE}` to use a different routing table for each TAP interface. Then, each VM would be connected to a different WireGuard peer. I'd then use `ip rule add iif ${WGINTERFACE} table ${WGTABLE}` so I can have a separate table for the WireGuard interface, as well.

Personally, I'd go with many interfaces with single peers. (One reason: I'm not sure how WireGuard behaves in the presence of multiple overlapping `AllowedIPs` settings. I suspect [routingtable.c](https://git.zx2c4.com/WireGuard/tree/src/routingtable.c) has the answer, maybe you could enlighten us. ;) Using standard routing enables the use of ECMP.)

But WireGuard is flexible enough to also allow single interfaces with many peers.

To be honest, there's also a case to be made for per-AllowedIPs routing tables, but:

1. I don't readily see how to implement that elegantly without breaking the interface, and
2. probably not too many people require this functionality.